### PR TITLE
Various Stability Fixes

### PR DESCRIPTION
**Pull request by cdalton713**
- Extension works with dev tools open
- Hide token once it's set
- Look for uncategorized + unapproved Amazon transactions
- Color Popup
- Auto redirect to transactions page
- No longer crashes when you have <50 orders

### CODE REVIEW

File: src/managers/ynab-manager/index.ts
Comment (KevinVillela) [inline src/managers/ynab-manager/index.ts:1]: Is this file necessary? I thought that just the `export class YnabManager` in ynab-manager.ts was enough.

File: package.json [hunk 1]
```diff
@@ -7,6 +7,7 @@
   "scripts": {
     "build": "webpack --config webpack/webpack.config.cjs",
     "build_dev": "webpack --config webpack/webpack_dev.config.cjs",
+    "dev": "webpack --config webpack/webpack_dev.config.cjs --watch",
```
Comment (KevinVillela) [inline package.json:10]: So we don't have to repeat the command, I think you can do `npm run build_dev -- --watch`

File: package.json [hunk 2]
```diff
@@ -16,9 +17,10 @@
     "@jest/globals": "^29.7.0",
     "@types/chrome": "^0.0.258",
     "@types/jest": "^29.5.12",
+    "@types/lodash": "^4.17.0",
     "copy-webpack-plugin": "^12.0.2",
-    "jest": "^29.7.0",
-    "jest-environment-jsdom": "^29.7.0",
+    "vitest": "^1.4.0",
```
Comment (KevinVillela) [inline package.json:22]: Just curious, why switch from jest to vitest? Does `npm test` still pass?

File: package.json [hunk 3]
```diff
@@ -28,6 +30,8 @@
   },
   "dependencies": {
     "linkedom": "^0.16.8",
-    "ynab": "^2.2.0"
+    "lodash": "^4.17.21",
```
Comment (KevinVillela) [inline package.json:33]: What is lodash for?

File: src/clients/ynab-client.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
+import * as ynab from "ynab";
+import * as statusAnd from "../utils/statusAnd";
+import {DateTime} from "luxon"
+
+export class YnabClient {
+  private accessToken: string;
+  private client: ynab.API
```
Comment (KevinVillela) [inline src/clients/ynab-client.ts:7]: I think these can be readonly.

File: src/clients/ynab-client.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    this.accessToken = token;
+    this.client = new ynab.API(token);
+  }
+
+  async getBudgets() {
+    const response = await this.client.budgets.getBudgets();
+    const budgets = response.data.budgets;
+
+    if (!budgets.length) {
+      throw statusAnd.error('API call to get budgets was successful, but returned 0 budgets!');
```
Comment (KevinVillela) [inline src/clients/ynab-client.ts:19]: `statusAnd`s aren't meant to be thrown, they are returned - kind of like C++'s [StatusOr](https://abseil.io/docs/cpp/guides/status).

File: src/clients/ynab-client.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+export class YnabClient {
+  private accessToken: string;
+  private client: ynab.API
+
+  constructor(token: string) {
+    this.accessToken = token;
+    this.client = new ynab.API(token);
+  }
+
+  async getBudgets() {
```
Comment (KevinVillela) [inline src/clients/ynab-client.ts:14]: Add return types to these methods plz!

File: src/clients/ynab-client.ts [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    return transactions;
+  }
+
+  async getUncategorizedTransactions(options: { budgetId: string, limitDays?: number }) {
+    const transactions = await this.client.transactions.getTransactionsByType(options.budgetId, 'uncategorized').then(r => r.data.transactions);
+    if (options.limitDays) {
+      return transactions.filter(t => DateTime.fromISO(t.date).diffNow('days').days >= -(options.limitDays!))
+    }
+    return transactions;
+  }
```
Comment (KevinVillela) [inline src/clients/ynab-client.ts:55]: Should these be private?

File: src/managers/amazon-manager/amazon-manager.ts [hunk 8]
```diff
@@ -0,0 +1,183 @@
+import {DOMParser} from 'linkedom/worker';
+import {getActiveTab, getHtml} from "../../utils/utils";
+import {FullTransaction, OrderInfo, Transaction} from "./types";
+import * as statusAnd from "../../utils/statusAnd"
+
+export class AmazonManager {
```
Comment (KevinVillela) [inline src/managers/amazon-manager/amazon-manager.ts:6]: Is there a reason to have this be a class vs just exported functions?

File: src/utils/utils.ts
Comment (KevinVillela) [inline src/utils/utils.ts:1]: I'm generally not a fan of files named `utils`. This file seems to deal mostly with the web page, so maybe html.ts or web_page.ts or something similar. If no name makes sense, we can also split this into multiple files.

File: src/managers/transaction-parser.ts [hunk 9]
```diff
@@ -0,0 +1,34 @@
+import * as ynab from "ynab";
+import {FullTransaction} from "./amazon-manager";
+
+
+export const parseTransactions = (options: {
```
Comment (KevinVillela) [inline src/managers/transaction-parser.ts:5]: I think this can be `export function`?

Comment (KevinVillela) [inline src/managers/transaction-parser.ts:5]: This isn't really parsing transactions, it's merging/joining them.

File: src/managers/transaction-parser.ts [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+  ynabAmazonTransactions: ynab.TransactionDetail[],
+  amazonTransactions: FullTransaction[]
+}) => {
```
Comment (KevinVillela) [inline src/managers/transaction-parser.ts:8]: Add return type
(throughout)

File: src/popup.ts [hunk 11]
```diff
@@ -2,30 +2,34 @@
...
-    await chrome.storage.local.set({"ynab_access_token": ynabAccessToken});
-    await chrome.runtime.sendMessage({ type: "run", ynabAccessToken, "num_orders": 50});
-    // The response will come in a different message (see addListener below).
+  const ynabAccessTokenField = (document.getElementById('ynab_access_token') as HTMLInputElement).value;
+  const storedAccessToken = await chrome.storage.local.get("ynab_access_token").then(r => r.ynab_access_token);
+
+  if ((ynabAccessTokenField !== storedAccessToken) && !ynabAccessTokenField.includes("*")) {
+    await chrome.storage.local.set({"ynab_access_token": ynabAccessTokenField});
+  }
+  await chrome.runtime.sendMessage({type: "run", ynabAccessToken: storedAccessToken, "numOrders": 50});
```
Comment (KevinVillela) [inline src/popup.ts:24]: Does this need to be set to ynabAccessTokenField if ynabAccessTokenField !== storedAccessToken?

File: src/utils/utils.ts [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+    } else {
+      return Promise.resolve(document.body.innerHTML);
+    }
+  } catch (e) {
+    console.log(e)
+    return Promise.resolve("")
+  }
+}
+
+export const getHtml = async (tabId: number) => {
```
Comment (KevinVillela) [inline src/utils/utils.ts:21]: These can be functions as well, instead of constants.

**Review by KevinVillela — CHANGES_REQUESTED** [review]
Wooo awesome, thank you for the changes!! I took a quick pass, I will try to run it to make sure everything still works well if I get a chance tomorrow.

Btw, just something to keep in mind, I'm hoping to add the option to use another extension to do the Amazon HTML parsing: https://github.com/KevinVillela/ynab/issues/1. No action required about that now, just FYI.

File: src/managers/ynab-manager/index.ts
Comment (cdalton713) [reply src/managers/ynab-manager/index.ts:1]: It's pretty standard for JS projects.  By having `index.ts` you encapsulate the manager into that one folder.  So, image you add in some new helper functions in a new file `/managers/ynab-manager/utils.ts`, or you colocate some unit tests, etc... none of that is exported from `/managers/ynab-manager`.

File: package.json (hunk 1 above)
Comment (cdalton713) [reply package.json:10]: Having a `npm run dev` command that works out of the box, without needing to add additional flags, is generally expected

File: package.json (hunk 2 above)
Comment (cdalton713) [reply package.json:22]: I gave up on Jest for TS projects long ago haha.  It randomly started throwing a ESM error while I was working on it and didn't have the energy to figure out why.  Vitest "just works" 99% of the time and is much better, IMO.  Tests also still pass.

File: package.json (hunk 3 above)
Comment (cdalton713) [reply package.json:33]: Using uniqby to deduplicate some transactions.

File: src/clients/ynab-client.ts (hunk 4 above)
Comment (cdalton713) [reply src/clients/ynab-client.ts:7]: Yeah, I'd recommend setting up eslint and prettier in the package.  Didn't pay much attention to those details during the refactor.

File: src/managers/transaction-parser.ts (hunk 9 above)
Comment (cdalton713) [reply src/managers/transaction-parser.ts:5]: These are arrow functions - you can read up on the differences but in this case its just stylistic and a more concise syntax.

File: src/managers/transaction-parser.ts (hunk 10 above)
Comment (cdalton713) [reply src/managers/transaction-parser.ts:8]: I hear you, but also keep in mind Typescript infers return types automatically

File: src/popup.ts (hunk 11 above)
Comment (cdalton713) [reply src/popup.ts:24]: This updates chrome's storage to the new value if it goes from **** > some new token

File: src/managers/amazon-manager/amazon-manager.ts (hunk 8 above)
Comment (cdalton713) [reply src/managers/amazon-manager/amazon-manager.ts:6]: It better sets up adding new features to the project since you can construct the class.  It also helps with naming, since there's a clear difference of where the classes's method is coming from vs a directly imported function.

File: src/clients/ynab-client.ts (hunk 7 above)
Comment (cdalton713) [reply src/clients/ynab-client.ts:55]: eh maybe, just left them public since there might be a use for them later

File: src/clients/ynab-client.ts (hunk 5 above)
Comment (cdalton713) [reply src/clients/ynab-client.ts:19]: I haven't come across this setup in node projects before, felt it was a bit confusing tbh.  up to you!

File: package.json (hunk 1 above)
Comment (KevinVillela) [reply package.json:10]: Yes, I'm saying that this line can be
"dev": "npm run build_dev -- --watch"

File: package.json (hunk 3 above)
Comment (KevinVillela) [reply package.json:33]: Where? Sorry if it's obvious and I'm not seeing it.

File: src/clients/ynab-client.ts (hunk 5 above)
Comment (KevinVillela) [reply src/clients/ynab-client.ts:19]: I am open to other ideas! The approach you're taking is to throw an exception, but I would say that exceptions should be used only for truly exceptional cases, and then we still have the need for cases that we need to surface a warning to the user because we couldn't complete the action.

File: src/clients/ynab-client.ts (hunk 7 above)
Comment (KevinVillela) [reply src/clients/ynab-client.ts:55]: If they need a use for later, then we can make them public at that time. But I think you're other comment is right, we should let a linter police this sort of thing.

File: src/managers/amazon-manager/amazon-manager.ts (hunk 8 above)
Comment (KevinVillela) [reply src/managers/amazon-manager/amazon-manager.ts:6]: > It better sets up adding new features to the project since you can construct the class

I don't know what this means, can you elaborate?

> It also helps with naming

The functions should be named pretty clearly, and the importing code can do something like `import * as ...` if it wants to namespace.

I should have written in the dev guide that I generally try to follow https://google.github.io/styleguide/tsguide.html#container-classes, sorry!

File: src/managers/transaction-parser.ts (hunk 9 above)
Comment (KevinVillela) [reply src/managers/transaction-parser.ts:5]: I would say it's confusing to define functions as constants, I see no downsides to using the standard `function` syntax, and I don't see how it's more concise? The alternative would be

`export function parseTransactions(options: {....}) { `
which is shorter.

File: src/managers/transaction-parser.ts (hunk 10 above)
Comment (KevinVillela) [reply src/managers/transaction-parser.ts:8]: Yes, this is not for the compiler, it's for the reader.

File: src/managers/ynab-manager/index.ts
Comment (KevinVillela) [reply src/managers/ynab-manager/index.ts:1]: Gotcha. I don't feel strongly about it TBH, so this is fine.

File: src/popup.ts (hunk 11 above)
Comment (KevinVillela) [reply src/popup.ts:24]: ? Yes, but I'm talking about L24, which sends a message with `storedAccessToken`. But won't `storedAccessToken` be out of date if L22 gets executed?

File: src/clients/ynab-client.ts [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+
+  async getUncategorizedTransactions(options: { budgetId: string, limitDays?: number }) {
+    const transactions = await this.client.transactions.getTransactionsByType(options.budgetId, 'uncategorized').then(r => r.data.transactions);
+    if (options.limitDays) {
+      return transactions.filter(t => DateTime.fromISO(t.date).diffNow('days').days >= -(options.limitDays!))
```
Comment (KevinVillela) [inline src/clients/ynab-client.ts:52]: Do we need luxon for this? It seems unnecessary to have to import an external library (what is the size impact? It doesn't appear to support tree shaking AFAICT). I think it can be done easily in native JavaScript, something like
https://stackoverflow.com/a/11252167 which would be 5 LOC instead of importing a whole dependency.

**Review by KevinVillela — COMMENTED** [review]
Yeesh very sorry for the slow review, the last few weeks have been crazy

**Comment by cdalton713** [comment]
Sorry man I just don't have the capacity to make all these small/debatable changes you're requesting. I just wanted to get the plugin to work for me and figured I'd share it. Feel free to pull and update it as you need, or drop the PR, nbd.

**Comment by KevinVillela** [comment]
Hey, yeah I totally understand. I'm kind of a stickler for code quality, for better or worse. I appreciate you sharing the code, and if I get some time I will take business logic changes and try to integrate them!

Another approach, btw, would be to split the PR into just what is necessary to get the extension to work vs the other refactorings - it's usually a good idea for PRs to be focused on one goal anyway.
